### ISSUE TITLE
`SyntaxNodeProvider` should keep track of removed items

### ISSUE THREAD

**Issue by jjonescz** [dotnet/roslyn#67633]
*Originally posted by @chsienki in https://github.com/dotnet/roslyn/pull/66992#issuecomment-1495168414.*

When we construct the filter table, we will modify any existing items, but don't carry over the removed ones. 

https://github.com/dotnet/roslyn/blob/d697090189c072943c2a3d315cb011a713a37a4a/src/Compilers/Core/Portable/SourceGeneration/Nodes/PredicateSyntaxStrategy.cs#L101-L104

This surfaces as a correctness problem when the removed items are at the end of the table, but it also surfaces as a perf issue when the removed items are in the middle of the table. 

<details>
<summary>Consider the following test</summary>

```csharp

        [Fact, WorkItem(61162, "https://github.com/dotnet/roslyn/issues/61162")]
        public void IncrementalGenerator_Collect_SyntaxProvider3()
        {
            var generator = new IncrementalGeneratorWrapper(new PipelineCallbackGenerator(static ctx =>
            {
                var invokedMethodsProvider = ctx.SyntaxProvider
                    .CreateSyntaxProvider(
                        static (node, _) => node is InvocationExpressionSyntax,
                        static (ctx, ct) => ctx.SemanticModel.GetSymbolInfo(ctx.Node, ct).Symbol?.Name ?? "(method not found)")
                    .Select((n, _) => n)
                    .WithTrackingName("Select");

                ctx.RegisterSourceOutput(invokedMethodsProvider, static (spc, invokedMethod) =>
                {
                    spc.AddSource(invokedMethod, "// " + invokedMethod);
                });
            }));
            var source1 = """
                System.Console.WriteLine();
                System.Console.ReadLine();
                """;

            var source2 = """
                class C {
                    public void M()
                    {
                        System.Console.Clear();
                        System.Console.Beep();
                    }
                }
                """;

            var parseOptions = TestOptions.RegularPreview;
            Compilation compilation = CreateCompilation(new[] { source1, source2 }, options: TestOptions.DebugExeThrowing, parseOptions: parseOptions);

            GeneratorDriver driver = CSharpGeneratorDriver.Create(new[] { generator }, parseOptions: parseOptions, driverOptions: new GeneratorDriverOptions(IncrementalGeneratorOutputKind.None, trackIncrementalGeneratorSteps: true));
            verify(ref driver, compilation, new[]
            {
                "// WriteLine",
                "// ReadLine",
                "// Clear",
                "// Beep"
            });

            // edit part of source 1
            replace(ref compilation, parseOptions, """
                System.Console.WriteLine();
                System.Console.Write(' ');
                """);

            verify(ref driver, compilation, new[]
            {
                "// WriteLine",
                "// Write",
                "// Clear",
                "// Beep"
            });

            Assert.Collection(driver.GetRunResult().Results[0].TrackedSteps["Select"],
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Modified, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason)
                );

            // remove second line of source 1
            replace(ref compilation, parseOptions, """
                System.Console.WriteLine();
                """);

            verify(ref driver, compilation, new[]
            {
                "// WriteLine",
                "// Clear",
                "// Beep"
            });

            Assert.Collection(driver.GetRunResult().Results[0].TrackedSteps["Select"],
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Removed, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason),
                (r) => Assert.Equal(IncrementalStepRunReason.Cached, r.Outputs.Single().Reason)
                );

            static void verify(ref GeneratorDriver driver, Compilation compilation, string[] generatedContent)
            {
                driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out var outputCompilation, out var generatorDiagnostics);
                outputCompilation.VerifyDiagnostics();
                generatorDiagnostics.Verify();
                var trees = driver.GetRunResult().GeneratedTrees;
                Assert.Equal(generatedContent.Length, trees.Length);
                for (int i = 0; i < generatedContent.Length; i++)
                {
                    AssertEx.EqualOrDiff(generatedContent[i], trees[i].ToString());
                }
            }

            static void replace(ref Compilation compilation, CSharpParseOptions parseOptions, string source)
            {
                compilation = compilation.ReplaceSyntaxTree(compilation.SyntaxTrees.First(), CSharpSyntaxTree.ParseText(source, parseOptions));
            }
        }
```
</details>

This will fail on the last collection assertion. The table is returning `[[WriteLine, Cached], [Clear, Modified], [Beep, Modified]]`. `Clear` and `Beep` weren't modified, and so their downstream transforms shouldn't run but they do. This occurs because they're replacing the 'missing' `Write` which isn't tracked as being removed.

Basically I think we're not correctly tracking removals in the `PredicateSyntaxStrategy`.
